Introduced interface Launcher and renamed old to JUnit5Launcher

jlink opened the pull request:
This was done to address yesterday's discussion that `Launcher` should be a top-level abstraction in module `launcher-api`.

Moreover, it further decouples clients from implementation.

sbrannen commented:
The Javadoc in several locations needs to be updated to reference the `Launcher` instead of the concrete `JUnit5Launcher`.

jlink commented:
Hopefully all locations found.

2016-01-20 13:52 GMT+01:00 Sam Brannen <notifications@github.com>:

> The Javadoc in several locations needs to be updated to reference the
> Launcher instead of the concrete JUnit5Launcher.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit5/pull/129#issuecomment-173195414>.
>

sbrannen commented on junit-launcher/src/main/java/org/junit/gen5/launcher/main/package-info.java:
```diff
@@ -1,8 +1,8 @@
 /**
- * The {@link org.junit.gen5.launcher.main.Launcher Launcher} class is the
+ * The {@link org.junit.gen5.launcher.main.JUnit5Launcher Launcher} class is the
```
Half missed this one. :wink:

marcphilipp commented:
I think separating the interface from the implementation is a good idea even though we will only provide a single implementation. 

However, I think the name `JUnit5Laucher` is misleading since it can be used to execute tests for any engine. How about `DefaultLauncher`?

sbrannen commented:
> However, I think the name `JUnit5Laucher` is misleading since it can be used to execute tests for any engine. How about `DefaultLauncher`?

I agree: `DefaultLauncher` or `StandardLauncher` would be more more suitable names.

jlink commented:
I also thing that the number 5 should mostly be restricted to
"junit5-engine". "DefaultLauncher" suggests that many exist, what about
just "JUnitLauncher"?

2016-01-20 16:33 GMT+01:00 Marc Philipp <notifications@github.com>:

> I think separating the interface from the implementation is a good idea
> even though we will only provide a single implementation.
>
> However, I think the name JUnit5Laucher is misleading since it can be
> used to execute tests for any engine. How about DefaultLauncher?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit5/pull/129#issuecomment-173239988>.
>

marcphilipp commented:
Well, we (or someone else) might create another one at some point, e.g. A `RemoteLaucher` proxy, or a new one that uses a different engine registration model.

jlink commented:
So, do you vote for DefaultLauncher?

I could do the change in the PR. Problem, with all the polishing going on,
rebasing becomes an increasing problem since the PR is all over the place.

2016-01-20 17:08 GMT+01:00 Marc Philipp <notifications@github.com>:

> Well, we (or someone else) might create another one at some point, e.g. A
> RemoteLaucher proxy, or a new one that uses a different engine
> registration model.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit5/pull/129#issuecomment-173252231>.
>

marcphilipp commented:
At present, yes. 😀

I wouldn't mind if you did the rename in master after merging this pull request.

jlink commented:
Since I fixed all know errors in the PR and just rebased it, feel free to
go ahead and accept it.

2016-01-20 17:18 GMT+01:00 Marc Philipp <notifications@github.com>:

> At present, yes. 😀
>
> I wouldn't mind if you did the rename in master after merging this pull
> request.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit5/pull/129#issuecomment-173255750>.
>

sbrannen commented:
`DefaultLauncher` sounds good.

Thanks for making the changes in the branch, @jlink.

I'll merge this into `master` this evening.

jlink commented:
Cheers.

Von meinem iPad gesendet

> Am 20.01.2016 um 18:31 schrieb Sam Brannen <notifications@github.com>:
> 
> DefaultLauncher sounds good.
> 
> Thanks for making the changes in the branch, @jlink.
> 
> I'll merge this into master this evening.
> 
> —
> Reply to this email directly or view it on GitHub.
>

sbrannen commented:
Merged into `master` in c3c93beab822d6ea83683d53bcc7456d74f2d59e.
